src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -22,46 +22,83 @@
 #include "ray/stats/metric_defs.h"
 
 namespace {
+/// The error message constructed from below methods is user-facing, so please avoid
+/// including too much implementation detail or internal information.
+void AddActorInfo(const ray::gcs::GcsActor *actor,
+                  ray::rpc::ActorDiedErrorContext *mutable_actor_died_error_ctx) {
+  if (actor == nullptr) {
+    return;
+  }
+  RAY_CHECK(mutable_actor_died_error_ctx != nullptr);
+  mutable_actor_died_error_ctx->set_owner_id(actor->GetOwnerID().Binary());
+  mutable_actor_died_error_ctx->set_owner_ip_address(
+      actor->GetOwnerAddress().ip_address());
+  mutable_actor_died_error_ctx->set_node_ip_address(actor->GetAddress().ip_address());
+  mutable_actor_died_error_ctx->set_pid(actor->GetActorTableData().pid());
+  mutable_actor_died_error_ctx->set_name(actor->GetName());
+  mutable_actor_died_error_ctx->set_ray_namespace(actor->GetRayNamespace());
+  mutable_actor_died_error_ctx->set_class_name(actor->GetActorTableData().class_name());
+  mutable_actor_died_error_ctx->set_actor_id(actor->GetActorID().Binary());
+}
+
 const ray::rpc::ActorDeathCause GenRuntimeEnvFailedCause(const std::string &error_msg) {
   ray::rpc::ActorDeathCause death_cause;
   death_cause.mutable_runtime_env_failed_context()->set_error_message(error_msg);
   return death_cause;
 }
 
-const ray::rpc::ActorDeathCause GenNodeDiedCause() {
+const ray::rpc::ActorDeathCause GenNodeDiedCause(const ray::gcs::GcsActor *actor,
+                                                 const std::string ip_address,
+                                                 const NodeID &node_id) {
   ray::rpc::ActorDeathCause death_cause;
-  death_cause.mutable_node_died_context()->set_error_message(
-      "The actor is dead because its node has died.");
+  auto actor_died_error_ctx = death_cause.mutable_actor_died_error_context();
+  AddActorInfo(actor, actor_died_error_ctx);
+  actor_died_error_ctx->set_error_message(absl::StrCat(
+      "(ip=", ip_address,
+      ") The actor is dead because its node has died. Node Id: ", node_id.Hex()));
   return death_cause;
 }
 
-const ray::rpc::ActorDeathCause GenWorkerDiedCause() {
+const ray::rpc::ActorDeathCause GenWorkerDiedCause(
+    const ray::gcs::GcsActor *actor, const std::string &ip_address,
+    const ray::rpc::WorkerExitType &disconnect_type) {
   ray::rpc::ActorDeathCause death_cause;
-  death_cause.mutable_worker_died_context()->set_error_message(
-      "The actor is dead because its worker process has died.");
+  auto actor_died_error_ctx = death_cause.mutable_actor_died_error_context();
+  AddActorInfo(actor, actor_died_error_ctx);
+  actor_died_error_ctx->set_error_message(absl::StrCat(
+      "(ip=", ip_address,
+      ") The actor is dead because its worker process has died. Worker exit type: ",
+      ray::rpc::WorkerExitType_Name(disconnect_type)));
```

Comment:
These metadata could be included separately outside the error message, but I prefer to just include it in the error message because otherwise, we will start having many optional fields. Imo, it is hard to understand code if there are many optional fields within the same schema.

Reply:
If we'd like to include this into `mutable_actor_died_error_context `, we can probably parse unset field in the python code.